JustSaying/AwsTools/QueueCreation/RegionResourceCache.cs

```diff
@@ -1,33 +1,31 @@
-using System.Collections.Generic;
-
-namespace JustSaying.AwsTools.QueueCreation
-{
-    public class RegionResourceCache<T> : Dictionary<string, Dictionary<string, T>>, IRegionResourceCache<T>
-    {
-        public T TryGetFromCache(string region, string key)
-        {
-            if (! ContainsKey(region))
-            {
-                return default(T);
-            }
-
-            var regionDict = this[region];
-            if (! regionDict.ContainsKey(key))
-            {
-                return default(T);
-            }
-
-            return regionDict[key];
-        }
-
-        public void AddToCache(string region, string key, T value)
-        {
-            if (!ContainsKey(region))
-            {
-                this[region] = new Dictionary<string, T>();
-            }
-            var regionDict = this[region];
-            regionDict[key] = value;
-        }
-    }
-}
\ No newline at end of file
+using System.Collections.Concurrent;
+
+namespace JustSaying.AwsTools.QueueCreation
+{
+    public sealed class RegionResourceCache<T>
+    {
+        private readonly ConcurrentDictionary<string, ConcurrentDictionary<string, T>> _regionsData
+            = new ConcurrentDictionary<string, ConcurrentDictionary<string, T>>();

```

Comment:
So `ID<string, ID<string, T>> d = new CD<string, ID<string, T>>();` (shorthand as on phone) and then just use CD again when adding items.

Reply:
We are now using `GetOrAdd`, which is not present on `IDictionary`.

Reply:
That’s true, but if it was using the interfaces it could’ve just used the simpler TryGetValue() and the rest of the code wouldn’t have needed to know the detail.

Meh, don’t have to change it now, but using the interfaces would have made it less verbose but still benefited from the concurrent-ness as AFAIK the interface methods leverage the concrete methods where the concurrentness is made explicit.

Reply:
I am finding lines 26-27 less verbose than the `if TryGetValue` equivalent. It's no important though, as concurrency on adding regions (not keys within regions) seems to be an edge case,